pkg/cmd/webhooks/forward.go

```diff
@@ -0,0 +1,253 @@
+				retries = 0
+
+				go func() {
+					for {
+						select {
+						case <-done:
+							return
+						case reply := <-replies:
+							err := conn.WriteJSON(reply)
+							if err != nil {
+								return
+							}
+						}
+					}
+				}()
+			}
+			return nil
+		},
+	}
+	cmd.Flags().StringVarP(&opts.EventType, "event", "E", "", "Name of the event type to forward")
+	cmd.Flags().StringVarP(&opts.Repo, "repo", "R", "", "Name of the repo where the webhook is installed")
+	cmd.Flags().IntVarP(&opts.Port, "port", "P", 0, "Local port to receive webhooks on")
+	cmd.Flags().StringVarP(&opts.Host, "host", "H", "", "Host address of GitHub API")
+	return &cmd
+}
+
+// createHook issues a request against the GitHub API to create a dev webhook
+func createHook(o *hookOptions) (string, error) {
+	httpClient, err := o.HttpClient()
+	if err != nil {
+		return "", err
+	}
+	apiClient := api.NewClientFromHTTP(httpClient)
+	path := fmt.Sprintf("repos/%s/hooks", o.Repo)
+	req := createHookRequest{
+		Name:   "cli",
+		Events: []string{o.EventType},
+		Active: true,
+		Config: hookConfig{
+			ContentType: "json",
+			InsecureSSL: "0",
+		},
+	}
+
+	reqBytes, err := json.Marshal(req)
+	if err != nil {
+		return "", err
+	}
+	var res createHookResponse
+	err = apiClient.REST(o.Host, "POST", path, bytes.NewReader(reqBytes), &res)
+	if err != nil {
+		return "", err
+	}
+	return res.WsURL, nil
+}
+
+// handleWebsocket connects to the websocket server and saves received events to a channel
+func handleWebsocket(token string, url *url.URL, opts *hookOptions, events chan wsEventReceived) (*websocket.Conn, error) {
+	fmt.Fprintf(opts.IO.Out, "Attempting to connect to webhooks server...\n")
+	h := make(http.Header)
+	h.Set("Authorization", token)
+
+	c, resp, err := websocket.DefaultDialer.Dial(url.String(), h)
+	if err != nil {
+		if resp != nil {
+			bts, _ := io.ReadAll(resp.Body)
+			err = fmt.Errorf("ws err %d - %s - %v", resp.StatusCode, bts, err)
+		}
+		return nil, err
+	}
+
+	defer c.Close()
+
+	for {
+		var ev wsEventReceived
+		err := c.ReadJSON(&ev)
+		if err != nil {
+			return nil, err
+		}
+		events <- ev
+	}
+
+	return c, nil
+}
+
+// forwardEvents forwards events to the server running on the local port specified by the user
+func forwardEvents(opts *hookOptions, events chan wsEventReceived, replies chan httpEventForward) error {
+	for ev := range events {
+		// TODO remove before merging
+		log.Printf("[LOG] received event with headers: %v \n", ev.Header)
+		if webhookRcvServerURL == "" {
+			webhookRcvServerURL = fmt.Sprintf("http://localhost:%d", opts.Port)
+		}
+
+		req, err := http.NewRequest(http.MethodPost, webhookRcvServerURL, bytes.NewReader(ev.Body))
+		if err != nil {
+			return err
+		}
+
+		for k := range ev.Header {
+			req.Header.Set(k, ev.Header.Get(k))
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return err
+		}
+
+		defer resp.Body.Close()
```

Comment:
This defer is also happening inside a potentially long-running for loop, which means it can lead to a memory leak. Might be worth extraction the for-loop block to a function so defers work as expected, or just repeat the close inside the error block and outside of it below.

Reply:
I took it out of the endless for loop.